compiler/next/lib/resolution/can-pass.cpp

```diff
@@ -413,65 +413,121 @@ CanPassResult::canConvertParamNarrowing(const QualifiedType& actualQT,
+  if (actual == formal) {
+    return passAsIs();
+  }
 
-  // The compiler considers many patterns of "subtyping" as things that require
-  // implicit conversions (they often require implicit conversions in the
-  // generated C).  However not all implicit conversions are created equal. Some
-  // of them are implementing subtyping.  Here we consider an implicit
-  // conversion to be implementing "subtyping" if, in an ideal implementation,
-  // the actual could be passed to a `const ref` argument of the formal type.
+  bool instantiates = false;
+  bool converts = false;
+  bool fails = false;
 
-  if (actualT->isNilType() && formalT->isAnyPtrType())
-    return convert(SUBTYPE);
+  ClassTypeDecorator actualNily = actual.toBorrowed();
+  ClassTypeDecorator formalNily = formal.toBorrowed();
 
+  ClassTypeDecorator actualMgmt = actual.removeNilable();
+  ClassTypeDecorator formalMgmt = formal.removeNilable();
 
-  if (actualT->isClassType() || formalT->isClassType()) {
-    // TODO: port canCoerceAsSubtype
-    assert(false && "not implemented yet");
+  // consider nilability.
+  if (actualNily != formalNily) {
+    if (formalNily.isUnknownNilability())
+      instantiates = true; // instantiating with passed nilability
+    else if (actualNily.isNonNilable() && formalNily.isNilable())
+      converts = true; // non-nil to nil conversion
+    else
+      fails = true; // all other nilability cases
   }
 
-  // TODO: implement c_ptr, c_array conversions
+  // consider management.
+  if (actualMgmt != formalMgmt) {
+    if (formalMgmt.isUnknownManagement())
+      instantiates = true; // instantiating with passed management
+    else if (formalMgmt.isBorrowed())
+      converts = true; // management can convert to borrowed
+    else
+      fails = true;
+  }
 
-  // Check for other class subtyping
-  // Class subtyping needs conversions in order to generate C code.
+  if (fails)
+    return fail();
 
-  return fail();
+  // all class conversions are subtype conversions
+  ConversionKind conversion = converts ? SUBTYPE : NONE;
+
+  return CanPassResult(/*passes*/ true,
+                       instantiates,
+                       /*promotes*/ false,
+                       conversion);
 }
 
-CanPassResult CanPassResult::canPass(const QualifiedType& actualQT,
-                                     const QualifiedType& formalQT) {
+CanPassResult CanPassResult::canPassClassTypes(const ClassType* actualCt,
+                                               const ClassType* formalCt) {
+  // owned Child -> owned Parent
+  // owned Child -> owned Parent?
+  // ditto borrowed, etc
+
+  // check decorators allow subtyping conversion
+  CanPassResult decResult = canPassDecorators(actualCt->decorator(),
+                                              formalCt->decorator());
+
+  if (actualCt->decorator().isManaged() &&
+      formalCt->decorator().isManaged() &&
+      actualCt->manager() != formalCt->manager()) {
```

Comment:
Let's talk about this separately. I don't think `managementStyle` is a good name for it because it's supposed to point to the record defining the management, and that doesn't seem a "style" thing. At the very least I'd like to rename it in a different PR since I already have another branch based on this PR.